src/System.Linq/src/System/Linq/OrderedEnumerable.cs

```diff
@@ -279,58 +251,43 @@ public TElement FirstOrDefault(Func<TElement, bool> predicate)
             }
         }
 
-        public TElement Last()
+        public bool TryGetLast(out TElement last)
         {
-            CachingComparer<TElement> comparer = GetComparer();
             using (IEnumerator<TElement> e = source.GetEnumerator())
             {
-                if (!e.MoveNext()) throw Error.NoElements();
-                TElement value = e.Current;
-                comparer.SetElement(value);
-                while (e.MoveNext())
+                if (!e.MoveNext())
                 {
-                    TElement x = e.Current;
-                    if (comparer.Compare(x, false) >= 0) value = x;
+                    last = default(TElement);
+                    return false;
                 }
-                return value;
-            }
-        }
 
-        public TElement LastOrDefault()
-        {
-            CachingComparer<TElement> comparer = GetComparer();
-            using (IEnumerator<TElement> e = source.GetEnumerator())
-            {
-                if (!e.MoveNext()) return default(TElement);
+                CachingComparer<TElement> comparer = GetComparer();
                 TElement value = e.Current;
                 comparer.SetElement(value);
                 while (e.MoveNext())
                 {
-                    TElement x = e.Current;
-                    if (comparer.Compare(x, false) >= 0) value = x;
+                    TElement current = e.Current;
+                    if (comparer.Compare(current, false) >= 0) value = current;
                 }
-                return value;
+                last = value;
+                return true;
             }
         }
 
-        public TElement Last(int minIdx, int maxIdx)
+        public bool TryGetLast(int minIdx, int maxIdx, out TElement last)
         {
             Buffer<TElement> buffer = new Buffer<TElement>(source);
             int count = buffer.count;
-            if (minIdx >= count) throw Error.NoElements();
-            if (maxIdx < count - 1) return GetEnumerableSorter().ElementAt(buffer.items, count, maxIdx);
+            if (minIdx >= count)
+            {
+                last = default(TElement);
+                return false;
+            }
+            if (maxIdx < count - 1) last = GetEnumerableSorter().ElementAt(buffer.items, count, maxIdx);
             // If we're here, we want the same results we would have got from
             // Last(), but we've already buffered our source.
-            return Last(buffer);
-        }
-
-        public TElement LastOrDefault(int minIdx, int maxIdx)
-        {
-            Buffer<TElement> buffer = new Buffer<TElement>(source);
-            int count = buffer.count;
-            if (minIdx >= count) return default(TElement);
-            if (maxIdx < count - 1) return GetEnumerableSorter().ElementAt(buffer.items, count, maxIdx);
-            return Last(buffer);
+            else last = Last(buffer);
```

Comment:
This could be:
```C#
last = maxIdx < count - 1 ?
    GetEnumerableSorter().ElementAt(buffer.items, count, maxIdx) :
    Last(buffer);
return true;
```